packages/ra-core/src/form/useApplyInputDefaultValues.ts

```diff
@@ -13,13 +13,19 @@ export const useApplyInputDefaultValues = (
 ) => {
     const { defaultValue, source } = props;
     const record = useRecordContext(props);
-    const { getValues, resetField } = useFormContext();
+    const {
+        getValues,
+        resetField,
+        getFieldState,
+        formState,
+    } = useFormContext();
     const recordValue = get(record, source);
     const formValue = get(getValues(), source);
+    const { isDirty } = getFieldState(source, formState);
 
     useEffect(() => {
         if (defaultValue == null) return;
-        if (formValue == null && recordValue == null) {
+        if (formValue == null && recordValue == null && !isDirty) {
```

Comment:
Not checking `isDirty` makes the field reinitialize to default value when using a `parse` function to set null for empty values (as now in the `<TextInput>` by default).